python/tvm/relay/frontend/sklearn.py

```diff
@@ -0,0 +1,154 @@
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=invalid-name, import-self, len-as-condition, unused-argument, too-many-lines
+# pylint: disable=import-outside-toplevel
+"""ONNX: Open Neural Network Exchange frontend for Relay."""
+import numpy as np
+import tvm
+from tvm.ir import IRModule
+
+from ... import nd as _nd
+from .. import analysis
+from .. import expr as _expr
+from .. import function as _function
+from .. import op as _op
+from .. import vision as _vision
+
+from ..function import Function
+from ..expr import Call, Let
+from ..expr import If, Tuple, TupleGetItem
+from ..expr import RefCreate, RefRead, RefWrite
+from ..expr_functor import ExprFunctor
+from ..adt import Match, Clause
+
+from .common import AttrCvt, Renamer, ExprTable
+from .common import get_relay_op, new_var, infer_shape, infer_channels
+from .common import infer_type, get_name
+from .common import infer_value as _infer_value
+from .common import infer_value_simulated as _infer_value_simulated
+
+
+def _SimpleImputer(op, inexpr, dshape, dtype, columns=None):
+    boolean_mask = _op.isnan(inexpr)
+    fill_col = _op.const(np.array(op.statistics_, dtype=dtype))
+    input_shape = _op.shape_of(inexpr)
+    reps = _op.take(input_shape, _op.const([0]))
+    reps = _op.concatenate([reps, _op.const([1])], axis=0)
+
+    fill_val = _op.tile(fill_col, reps=reps)
+    indices =_op.const(np.arange(len(op.statistics_)))
+    fill_val = _op.take(fill_val, indices=indices, axis=1)
+
+    ret = _op.where(boolean_mask,
+                    fill_val,
+                    inexpr)
+    
+    return ret
+
+def _RobustImputer(op, inexpr, dshape, dtype, columns=None):
+    if columns: 
+        column_indices = _op.const(columns)
+        inexpr = _op.take(inexpr, indices=column_indices, axis=1)
+
+    if not op.mask_function:
+        inf_mask = _op.isinf(inexpr)
+        nan_val = _op.full_like(inexpr, _op.const(np.array(np.nan, dtype=dtype)))
+        inexpr = _op.where(inf_mask, nan_val, inexpr) 
+    ret = _SimpleImputer(op.simple_imputer_, inexpr, dshape, dtype, columns)
+
+    return ret 
+    
+def _ThresholdOneHotEncoder(op, inexpr, dshape, dtype, columns=None):
+    if columns: 
+        column_indices = _op.const(columns)
+        inexpr = _op.take(inexpr, indices=column_indices, axis=1)
+
+    num_cat = len(op.categories_)
+    cols = _op.split(inexpr, num_cat, axis=1)
+
+    out = [] 
+    for i in range(num_cat):
```

Comment:
Can we use the TVM `one_hot` operator instead? https://tvm.apache.org/docs/api/python/relay/index.html#tvm.relay.one_hot

Reply:
The one hot operator can't be used in this scenario because the one-hot encoder depends on the "categories" attribute from the trained model. Relay's one hot operator will encoder all values to indices randomly(or at least different from the indices defined by this attribute).

Reply:
```
indices = [0, 1, 2]

relay.one_hot(indices, depth=3) =
    [[1, 0, 0],
     [0, 1, 0],
     [0, 0, 1]]
```
For TVM, a item with a value of 0 will always produce a one hot vector of length `depth` where index 0 is "on". So `relay.one_hot([0], 3)` would give `[1, 0, 0]`.

It looks like for ThresholdOneHotEncoder, the position of the item in the `categories` list is what determines what index is on. Is that true? So if categories was [0, 1, 2], then it would be equivalent to TVM. But categories could be [2, 0, 1] which would mean an input `[0]` would become `[0, 1, 0]`?